user/config/config_gotls.go

```diff
@@ -311,8 +305,18 @@ func (gc *GoTLSConfig) ReadTable() (*gosym.Table, error) {
 		return nil, fmt.Errorf("could not find magic number in %s ", gc.Path)
 	}
 	tableData = tableData[pclntabIndex:]
-
-	addr := gc.goElf.Section(".text").Addr
+	var addr uint64
+	{
+		// get textStart from pclntable
+		// please see https://go-review.googlesource.com/c/go/+/366695
+		// tableData
+		ptrSize := uint32(tableData[7])
+		if ptrSize == 4 {
+			addr = uint64(binary.LittleEndian.Uint32(tableData[8+2*ptrSize:]))
+		} else {
+			addr = binary.LittleEndian.Uint64(tableData[8+2*ptrSize:])
```

Comment:
What is the relationship between `gc.goElf.Section(".text").Addr `  and this ?

Reply:
Well, I don't know the relationship between `.text` section address and this address ("runtime.text"), I think in some cases, there are same address, but no more. I just know relationship between"runtime.text" address and function address.

Package `debug/gosym` implement parse `pclntab`, in this code, we'll see [go pr 366695](https://go-review.googlesource.com/c/go/+/366695), this commit show `textStart` should be got from `pclntab` not address of `.text` section, and this address only use in golang version >= 1.18.

And I found something in golang link, which create `pclntab`, in Function [`ld/pcln.go:generatePCHeader`](https://github.com/golang/go/blob/master/src/cmd/link/internal/ld/pcln.go#L230), we will see `textStartOff := int64(8 + 2*ctxt.Arch.PtrSize)` and set `ldr.Lookup("runtime.text", 0)` to this offset, this function will called by [`link.pclntab`](https://github.com/golang/go/blob/master/src/cmd/link/internal/ld/pcln.go#L757) that  initializes the pclntab symbol.
And we see [`generateFunctab`](https://github.com/golang/go/blob/master/src/cmd/link/internal/ld/pcln.go#L535), this function also called by `link.pclntab`, this function call function [`writePCToFunc`](https://github.com/golang/go/blob/master/src/cmd/link/internal/ld/pcln.go#L604) and [`writeFuncs`](https://github.com/golang/go/blob/master/src/cmd/link/internal/ld/pcln.go#L625).

This two function all use this code when set function offset to `pclntab`. 
```golang
textStart := ldr.SymValue(ldr.Lookup("runtime.text", 0))

off := ldr.SymValue(s) - textStart
```

and in `debug/gosym`, we'll see [this code](https://github.com/golang/go/blob/master/src/debug/gosym/pclntab.go#L416) when parse function offset and calculate function address from `pclntab`: 
```golang
	u := f.uint(f.functab[2*i*f.sz:])
	if f.version >= ver118 {
		u += f.textStart
	}
```

Reply:
Thank you very much for your explanation, I wasn't very familiar with this part before, I learned a lot.